Add doc for config file (#6392)
ocaml/dune: doc/dune-files.rst

Stanza Reference
****************

.. _config:

``config``
==========

This file is used to set global configuration of Dune (applicable across
projects and workspaces).

The configuration file is normally ``~/.config/dune/config`` on Unix systems and
``%LOCALAPPDATA%/dune/config`` on Windows. However, it is possible to specify an

> alternative configuration file with the ``--config-file`` option.

Author: Yes, fixed.

Command-line flags take precedence over the contents of the ``config`` file.

> If ``--no-config`` or ``-p`` is passed, Dune will not read this file.

Author: Yes, fixed.

It can contain the following stanzas:

.. _display:

``display``
-----------

Specify the amount of Dune’s verbosity.

.. code:: scheme

    (display <setting>)

where ``<setting>`` is one of:

- ``progress``, Dune shows and updates a status line as build goals are being

>   completed (this is the default).

Reviewer: ```suggestion
  completed. This is the default value.
```

Author: Fixed.

> - ``verbose`` prints the full command lines of programs being executed by Dune,

Reviewer 2: ```suggestion
- ``verbose`` prints the full command lines of programs executed by Dune,
```
(looking to tighten syntax a little, but it's fine the way it is if you prefer)

  with some colors to help differentiate programs.

- ``short`` prints one line per executed command, with the binary name on
  the left and the reason it is being executed for on the right.

- ``quiet`` only display errors.

.. _concurrency:

``concurrency``
---------------

> Number of cores Dune is allowed to utilize.

Reviewer: So "cores" might not be the correct thing to say here. Perhaps:
```
Maximum number of concurrent jobs Dune is allowed to have.
```

Author: Fixed (both).

.. code:: scheme

    (concurrency <setting>)

where ``<setting>`` is one of:

> - ``auto``, auto-detect maximum number of cores.

Reviewer: Add: `This is the default value.`

Author: Fixed.

> - ``<number>``, a positive integer specifying the maximum number of cores Dune

Author: Fixed.

  may use simultaneously.

.. _terminal-persistence:

``terminal-persistence``
------------------------

Specifies how Dune handles the terminal when a rebuild is triggered in watch mode.

.. code:: scheme

    (terminal-persistence <setting>)

where ``<setting>`` is one of:

- ``preserve`` does not clear the terminal screen beteween rebuilds.

- ``clear-on-rebuild`` clears the terminal screen between rebuilds.

- ``clear-on-rebuild-and-flush-history`` clears the terminal between rebuilds, and
  it also deletes everything in the scrollback buffer.

.. _sandboxing_preference:

``sandboxing_preference``
-------------------------

The preferred sandboxing setting. Individual rules may specify different
preferences. Dune will try to utilize a setting satisfying both conditions.

.. code:: scheme

    (sandboxing_preference <setting> <setting> ...)

where each ``<setting>`` can be one of:

> - ``none``, no sandboxing allowed

Author: Fixed.

> - ``symlink``, using symbolic links

Reviewer: ```suggestion
- ``symlink`` uses symbolic links for sandboxing.
```
Also is this the default? If so, mention it.

Author: Fixed (both).

> - ``copy``, using file copies

Author: Fixed.

> - ``hardlink``, using hard links

Author: Fixed.

.. _cache:

``cache``
---------

> Specifies whether to utilize the Dune cache.

Author: Fixed.

.. code:: scheme

    (cache <setting>)

where ``<setting>`` is one of:

- ``enabled`` enables Dune cache.

- ``disabled`` disables Dune cache.

.. _cache_check_probability:

``cache-check-probability``
---------------------------

While the main purpose of Dune cache is to speed up build times, it can also be
used to check build reproducibility. It is possible to enable a probabilistic
check, in which Dune will re-execute randomly chosen build rules and compare
their results with those stored in the cache. If the results differ, the rule is
not reproducible, and Dune will print out a corresponding warning.

.. code:: scheme

    (cache-check-probability <number>)

where ``<number>`` is a floating-point number between 0 and 1 (inclusive). 0
means never to check for reproducibility, and 1 means to always perform the
check.

.. _cache_storage_mode:

``cache-storage-mode``
----------------------

> Specify the mechanism used by the Dune cache storage layer.

Author: Fixed.

.. code:: scheme

    (cache-storage-mode <setting>)

where ``<setting>`` is one of:

> - ``auto``, Dune decides the best mechanism to use.

Author: Fixed.

> - ``hardlink``, using hard links.

Reviewer: Also we should mention that if a user is storing the cache in a different disk partition to the one where the build takes place that this mode will not work and they should instead use copy.

> - ``hardlink``, using hard links.

Author: Fixed.

> - ``copy``, using file copies; this is less efficient than using hard links.

Reviewer: ```suggestion
- ``copy`` copies entries to the cache. This is less efficient than using hard links.
```

Author: Fixed.

.. _action_stdout_on_success:

``action_stdout_on_success``
----------------------------

Specifies how Dune should handle the standard output of actions when they succeed.
This can be used to reduce the noise of large builds.

.. code:: scheme

    (action_stdout_on_success <setting>)

where ``<setting>`` is one of:

> - ``print`` prints the output on the terminal (this is the default).

Reviewer: ```suggestion
- ``print`` prints the output on the terminal. This is the default.
```

Author: Fixed.

- ``swallow`` ignores the output and does not print it on the terminal.

- ``must-be-empty`` enforces that the output should be empty. If it is not, Dune will fail.

.. _action_stderr_on_success:

``action_stderr_on_success``
----------------------------

Same as :ref:`action_stdout_on_success`, but applies to standard error instead
of standard output.

.. _dune-project:

> ``dune-project``

Reviewer: Let's not do this as it makes the headers weird.

Author: Fixed.

Stanza Reference
****************

.. _config:

config
======

This file is used to set global configuration of Dune (applicable across
projects and workspaces).

The configuration file is normally ``~/.config/dune/config`` on Unix systems and
``%LOCALAPPDATA%/dune/config`` on Windows. However, it is possible to specify an
alternative configuration file with the ``--config-file`` option.

> Command-line flags take precedence over the contents of the config file.

Reviewer 2: ```suggestion
Command-line flags take precedence over the contents of the ``config`` file.
```

If ``--no-config`` or ``-p`` is passed, Dune will not read this file.

It can contain the following stanzas:

.. _display:

display
-------

> Specify the amount of verbosity of Dune.

Reviewer 2: ```suggestion
Specify the amount of Dune’s verbosity.
```

?? Maybe??

.. code:: scheme

    (display <setting>)

where ``<setting>`` is one of:

> - ``progress`` Dune shows and update a status line as build goals are being

Reviewer 2: ```suggestion
- ``progress``, Dune shows and updates a status line as build goals are being
```

  completed (this is the default).

> - ``verbose`` print the full command lines of programs being executed by Dune,

Reviewer 2: ```suggestion
- ``verbose`` prints the full command lines of programs being executed by Dune,
```

  with some colors to help differentiate programs.

> - ``short`` print one line per command being executed, with the binary name on

Reviewer 2: ```suggestion
- ``short`` prints one line per executed command, with the binary name on
```

>   the left and the reason it is being executed for on the right.

Reviewer 2: ```suggestion
  the left and the reason for execution on the right.
```

?? Is “execution” used in this way?

- ``quiet`` only display errors.

.. _concurrency:

concurrency
-----------

Number of cores Dune is allowed to utilize.

.. code:: scheme

    (concurrency <setting>)

where ``<setting>`` is one of:

- ``auto``, auto-detect maximum number of cores.

- ``<number>``, a positive integer specifying the maximum number of cores Dune
  may use simultaneously.

.. _terminal-persistence:

terminal-persistence
--------------------

> Specify how Dune handles the terminal when a rebuild is triggered in watch mode.

Reviewer 2: ```suggestion
Specifies how Dune handles the terminal when a rebuild is triggered in watch mode.
```

.. code:: scheme

    (terminal-persistence <setting>)

where ``<setting>`` is one of:

> - ``preserve`` do not clear terminal screen beteween rebuilds.

Reviewer 2: ```suggestion
- ``preserve`` does not clear the terminal screen beteween rebuilds.
```

> - ``clear-on-rebuild`` clear terminal screen between rebuilds.

Reviewer 2: ```suggestion
- ``clear-on-rebuild`` clears the terminal screen between rebuilds.
```

> - ``clear-on-rebuild-and-flush-history`` clear terminal between rebuilds, and

Reviewer 2: ```suggestion
- ``clear-on-rebuild-and-flush-history`` clears the terminal between rebuilds, and
```

>   also delete everything in the scrollback buffer.

Reviewer 2: ```suggestion
it also deletes everything in the scrollback buffer.
```

.. _sandboxing_preference:

sandboxing_preference
---------------------

> Your preferred sandboxing setting. Individual rules may specify different

Reviewer 2: ```suggestion
The preferred sandboxing setting. Individual rules may specify different
```

> preferences, Dune will try to utilize a setting satisfying both conditions.

Reviewer 2: ```suggestion
preferences. Dune will try to utilize a setting satisfying both conditions.
```

.. code:: scheme

    (sandboxing_preference <setting> <setting> ...)

where each ``<setting>`` can be one of:

> - ``none``, no sandboxing allowed.

Reviewer 2: ```suggestion
- ``none``, no sandboxing allowed
```

> - ``symlink``, using symbolic links.

Reviewer 2: ```suggestion
- ``symlink``, using symbolic links
```

> - ``copy``, using file copies.

Reviewer 2: ```suggestion
- ``copy``, using file copies
```

> - ``hardlink``, using hard links.

Reviewer 2: ```suggestion
- ``hardlink``, using hard links
```

.. _cache:

cache
-----

> Specify whether to utilize the Dune cache.

Reviewer 2: ```suggestion
Specifies whether to utilize the Dune cache.
```

.. code:: scheme

    (cache <setting>)

where ``<setting>`` is one of:

> - ``enabled``, enable Dune cache.

Reviewer 2: ```suggestion
- ``enabled`` enables Dune cache.
```

> - ``disabled``, disble Dune cache.

Reviewer 2: ```suggestion
- ``disabled`` disables Dune cache.
```

.. _cache_check_probability:

cache-check-probability
-----------------------

While the main purpose of Dune cache is to speed up build times, it can also be
used to check build reproducibility. It is possible to enable a probabilistic

> check in which Dune will re-execute randomly chosen build rules and compare

Reviewer 2: ```suggestion
check, in which Dune will re-execute randomly chosen build rules and compare
```

their results with those stored in the cache. If the results differ, the rule is

> not reproducible and Dune will print out a corresponding warning.

Reviewer 2: ```suggestion
not reproducible, and Dune will print out a corresponding warning.
```

.. code:: scheme

    (cache-check-probability <number>)

where ``<number>`` is a floating-point number between 0 and 1 (inclusive). 0
means never to check for reproducibility, and 1 means to always perform the
check.

.. _cache_storage_mode:

cache-storage-mode
------------------

Specify the mechanism used by the Dune cache storage layer.

.. code:: scheme

    (cache-storage-mode <setting>)

where ``<setting>`` is one of:

- ``auto``, Dune decides the best mechanism to use.

> - ``hardlink``, use hard links.

Reviewer 2: ```suggestion
- ``hardlink``, using hard links.
```
To match the formatting in the previous list

> - ``copy``, use file copies; this is less efficient than using hard links.

Reviewer 2: ```suggestion
- ``copy``, using file copies; this is less efficient than using hard links.
```

.. _action_stdout_on_success:

action_stdout_on_success
------------------------

> Specify how Dune should handle the standard output of actions when they succeed.

Reviewer 2: ```suggestion
Specifies how Dune should handle the standard output of actions when they succeed.
```

> This cane be used to reduce the noise of large builds.

Reviewer 2: ```suggestion
This can be used to reduce the noise of large builds.
```

.. code:: scheme

    (action_stdout_on_success <setting>)

where ``<setting>`` is one of:

> - ``print``, print the output on the terminal (this is the default).

Reviewer 2: ```suggestion
- ``print`` prints the output on the terminal (this is the default).
```

> - ``swallow``, ignore the output and do not print it on the terminal.

Reviewer 2: ```suggestion
- ``swallow`` ignores the output and does not print it on the terminal.
```

> - ``must-be-empty``, enforce that the output should be empty. If it is not, Dune will fail.

Reviewer 2: ```suggestion
- ``must-be-empty`` enforces that the output should be empty. If it is not, Dune will fail.
```

.. _action_stderr_on_success:

action_stderr_on_success
------------------------

> Same as :ref:`action_stdout_on_success`, but applying to standard error instead

Reviewer 2: ```suggestion
Same as :ref:`action_stdout_on_success`, but applies to standard error instead
```

.. _using:

using
-----

The language of configuration files read by Dune can be extended to support

> additional stanzas (eg., ``menhir``, ``coq.theory``, ``mdx``). This is done by

Reviewer 2: (e.g., …)
also  `.` after the `e`

For rare cases where this is not what you want, you can force Dune to use a
different build contexts for Merlin by adding the field ``(merlin)`` to this
context.

``config`` stanzas
------------------

> Moreover, starting in Dune 3.0, any of the stanzas from the :ref:`config` file

Reviewer: ```suggestion
Starting in Dune 3.0, any of the stanzas from the :ref:`config` file
```
and rewrap

Author: Thanks, fixed.

can be used in the ``dune-workspace`` file. In this case, the configuration
stanza will only affect the current workspace.

.. _config:

config
======

> This file is used to set global configuration of Dune (applicable across

Reviewer: ```suggestion
This file is used to set the global configuration of Dune (applicable across
```

> This file is used to set global configuration of Dune (applicable across

Reviewer: ```suggestion
This file is used to set global configuration of Dune which is applicable across
```
and remove the brackets.

Author: Fixed.

projects and workspaces).

The configuration file is normally ``~/.config/dune/config`` on Unix systems and
``%LOCALAPPDATA%/dune/config`` on Windows. However, for most Dune commands it is
possible to specify an alternative configuration file with the ``--config-file``
option. Command-line flags take precedence over the contents of the ``config``
file.  If ``--no-config`` or ``-p`` is passed, Dune will not read this file.

> It can contain the following stanzas:

Reviewer: ```suggestion
The ``config`` file can contain the following stanzas:
```

Author: Fixed.

.. _display:

display
-------

Specify the amount of Dune’s verbosity.

.. code:: scheme

    (display <setting>)

where ``<setting>`` is one of:

- ``progress``, Dune shows and updates a status line as build goals are being
  completed. This is the default value.

- ``verbose`` prints the full command lines of programs being executed by Dune,
  with some colors to help differentiate programs.

- ``short`` prints a line for each program executed with the binary name on the
  left and the targets of the action on the right.

- ``quiet`` only display errors.

.. _concurrency:

concurrency
-----------

Maximum number of concurrent jobs Dune is allowed to have.

.. code:: scheme

    (concurrency <setting>)

where ``<setting>`` is one of:

- ``auto``, auto-detect maximum number of cores. This is the default value.

- ``<number>``, a positive integer specifying the maximum number of jobs Dune
  may use simultaneously.

> .. _terminal-persistence:

Reviewer: This appears to be made up. We should have called this jobs.
